### PR TITLE
[Flax] adding support for batch norm layers

### PR DESCRIPTION
**Pull request by Shubhamai**
# What does this PR do?

Adding support for batch norm layers in flax. Several models like ResNets, MobileNets have BatchNorm layers which (for now) cannot be ported to HuggingFace Flax due to Flax requiring an extra `batch_stats` key along with `params` to perform training and inference. 

The changes here are tested locally with my work-in-progress resnet PR [https://github.com/huggingface/transformers/pull/21472](https://github.com/huggingface/transformers/pull/21472), ~~though more testing is needed.~~ LGTM :rocket: 

### Design Choices
_Mentioning a few design decisions and some notes to ease the review process_ :)

- To detect whether the flax model contains batch norm layers, we use a simple if statement 
        `if "params" in flax_model.params and "batch_stats" in flax_model.params`
        An alternative could have been adding a `contains_batchnorm` attribute to `FlaxPreTrainedModel` class but I decided not to go with it since found the first option simpler and kept no. of class attributes as low as possible unless strictly needed. 
- `num_batches_tracked` is not transferred from pt to flax weight conversion and are newly initialized when doing flax to pt weight conversion.
- ~~`pt_flax_cross_test` in `test_modeling_flax_common.py` currently does not support comparing the weights between  `batch_stats` and PyTorch equivalent BatchNorm mean and var. I am still currently looking into supporting it.~~ 
        It does now :)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

- Flax: @sanchit-gandhi

### CODE REVIEW

**Comment by Shubhamai** [comment]
Hi @sanchit-gandhi, the PR is now ready for your review. I'm sorry if my previous ping was confusing since I mentioned it when the PR was in WIP. Next time, I will make sure to ping you only when it's ready for review. Thank you so much for taking the time to review the PR.

[Flax Convnext PR](https://github.com/huggingface/transformers/pull/21485) is also ready for your review.

**Comment by Shubhamai** [comment]
HI, @sgugger  I have been trying to mention @sanchit-gandhi but unfortunately have not received a response. I was hoping you could assist me in finding the right person to reach out to for a review. Any help would be greatly appreciated.

**Comment by sgugger** [comment]
The right person is @sanchit-gandhi

File: src/transformers/modeling_flax_pytorch_utils.py
Comment (sanchit-gandhi) [inline src/transformers/modeling_flax_pytorch_utils.py:133]: Could we not always set:
```python
flax_model_params = flax_model.params["params"]
random_flax_state_dict = flatten_dict(flax_model_params)
```
And then check if `"batch_stats"` is in the flax params dict:
```python
if "batch_stats" in flax_model.params:
    flax_batch_stats = flatten_dict(flax_model.params["batch_stats"])
    random_flax_state_dict.update(flax_batch_stats)
```

Just to simplify the `if/else` logic a bit. If I understand correctly then we'll always have `"params"` in the params dict, and only `"batch_stats"` if using batch norm (as per https://flax.readthedocs.io/en/latest/guides/batch_norm.html#the-batch-stats-collection).

File: src/transformers/modeling_flax_pytorch_utils.py
Comment (sanchit-gandhi) [inline src/transformers/modeling_flax_pytorch_utils.py:178]: Again here I wonder if it would be cleaner to always set:
```python
flax_state_dict[flax_key] = jnp.asarray(flax_tensor)
```
And then check if `"batch_stats"` is in `flax_model.params`:
```python
if "batch_stats" in flax_model.params:
    if "mean" in flax_key[-1]:
        ...
```

**Review by sanchit-gandhi — COMMENTED** [review]
Really clean PR @Shubhami, nice one! Super sorry for the late reply here - please accept my apologies!

I fully back your design decision to use `if/else` statements to check the Flax params dict for `"batch_norm"` entries rather than adding another attribute to `FlaxPreTrainedModel` 👌

My only question is whether we can perhaps simplify the `if/else` logic by one degree? See comments below!

File: src/transformers/modeling_flax_pytorch_utils.py
Comment (Shubhamai) [reply src/transformers/modeling_flax_pytorch_utils.py]: Completely agree, made the changes in the most recent commit.

File: src/transformers/modeling_flax_pytorch_utils.py
Comment (Shubhamai) [reply src/transformers/modeling_flax_pytorch_utils.py]: I added params checking in `if/else` just to extra make sure we have batch norm layers, but I agree it can be simplified, made the changes in the most recent commit.

**Review by sanchit-gandhi — APPROVED** [review]
Very clean PR @Shubhamai! Thanks for making the changes to the `if/else` checks - it's super easy to follow the logic both with/without batch norm.

My guess is that we'll implicitly get a test for these changes once we've finished Flax ResNet (i.e. once we have a model that has batch norm layers)?

Requesting a final review from @sgugger for a core maintainer's approval! 🤗

**Comment by Shubhamai** [comment]
Cool! Such a pleasure to work with :hugs:  team & transformers. Along with ResNets, I have tested this PR on flax implementation of RegNet and EfficientNet locally. Both of which are nearly complete and will make PR once this get's merged.

**Comment by sanchit-gandhi** [comment]
Super fun working with you too @Shubhamai! Thank you for your contribution and high quality PR! Very excited to see the follow-ups with ResNet, RegNet and EfficientNet, these will be very nice additions to the library. Feel free to ping me when they're in a review ready state and I'll take a look as quickly as possible! Likewise if you have any Flax questions / queries I'm more than happy to help!

**Review by sgugger — APPROVED** [review]
Thanks for your work on this!

**Pull request merged**
